stdlib/runtime/exception.gr

```diff
@@ -82,7 +82,7 @@ export exception Overflow
 export exception NumberNotIntlike
 export exception NumberNotRational
 export exception MatchFailure
-export exception AssertionError
+export exception AssertionError(String) // <- since we can't malloc in this module, we put the full error message in the Exception object
```

Comment:
I don't think we need this comment here since I don't think there's any expectation that it should be any other way:
```suggestion
export exception AssertionError(String)
```

Reply:
I disagree. I think that it merits some sort of comment, since (a) it explains what the `String` here is, (b) a much more natural choice of Exception object would be `AssertionError(String, Number)` (containing the location information of the failure), so this comment helps explain why the exception is not formatted like that, and (c) I don't think there's a compelling reason to not err on the side of having too much explanation.

Reply:
I disagree with (c) because of the "made you think" factor. If you already understand the constraints of this module and how assertions work this makes sense, but otherwise it's confusing because this doesn't explain how the AssertionError gets thrown or that the information is automatically populated. Without the comment, the thought is just "this assertion error has a message with it", but with the comment and no knowledge of how `assert` works, you might question "Why can't we malloc? Why does it matter that we can't malloc, and how would this be different? What do you mean by 'we' put the full error message in there? Does this mean users need to provide a full error message? What does a 'full' error message entail? Does that mean this is a hack of some kind?"

You can have a comment here, but you'd have to explain a bit more:
```suggestion
/*
 * `assert <condition>` is essentially syntactic sugar for:
 * `if (!<condition>) throw AssertionError(<message>)`
 * where <message> is a string injected by the compiler, containing file and line information.
 * This would be better as `AssertionError(String, Number)` to allow the exception printer in
 * this file to construct the error, but constructing a string in this module is impractical as we
 * don't have access to memory allocation since malloc.gr depends on this module.
 */
export exception AssertionError(String)
```

I think this information is useful, but none of it is really a "need to know" for this module IMO. I think it could live somewhere else, though.